spec/github_helper_spec.rb

```diff
@@ -234,14 +234,14 @@ def get_milestone(milestone_name:)
     end
 
     def create_milestone(need_submission:, milestone_duration:, days_code_freeze:)
+      days_until_submission = need_submission ? (days_code_freeze - 3) : milestone_duration
```

Comment:
Since this file is focused on testing the helper, and the helper doesn't have knowledge of the concept of `need_submission` anymore, I think we should adapt the `github_helper_spec` accordingly, to write the `it` tests based on the new parameters (i.e. don't test "has correct dates (with|without) submission" cases, but test things like "computes the correct dates for (a standard case | dates crossing DST boundaries | inconsistent offsets | …)".

Also begs the question what the behavior should do if we provide inconsistent inputs (invalid date, negative values for `days_until_*`, `days_until_release < days_until_submission`, …?).
Maybe it's ok to let that to the caller's responsibility and not check (i.e. keep the current behavior), at least for now; but adding tests to the action like here would also be the best time to raise this kind of questions (after all, if we try to follow in the footsteps of TDD, we should write the tests to describe the behaviors and use cases we want first, and then we implement them to match, not the other way around).

Reply:
> Also begs the question what the behavior should do if we provide inconsistent inputs (invalid date, negative values for days_until_*, days_until_release < days_until_submission, …?). [.....]

That's a question that I keep asking myself 🤔 💭 

Using as an example a case where I have more days from submission than from release (i.e `days_until_submission: 14`, `days_until_release: 3`), The submission date will be waaay after the release date. In this case, should we raise a `FastlaneError`?

```ruby
# Milestone with due_date = '2022-10-20' will generate
Code freeze: October 20, 2022
App Store submission: November 03, 2022
Release: October 23, 2022
```

--------------------------

Also, in the cases that we have negative values for these same parameters  (i.e `days_until_submission: -14`, `days_until_release: -14`), the Ruby DateApi at `param.to_datetime.next_day(..)` just subtracts days from the `due_date`, although is allowed, it is another inconsistency. So, negative values should be allowed on these parameters, in the first place?

```ruby
# Milestone with due_date = '2022-10-20' will generate
Code freeze: October 20, 2022
App Store submission: October 06, 2022
Release: October 06, 2022
```
-----------------------------

Also, assuming that when we refer to `Release`, we are talking about the date when this version will be available to our users, and by `AppStore Submission`, is the date we submit the code to be reviewed and alpha/beta tested. 

What is the expected behavior of the dates when creating milestones?

- The `Code Freeze` date should always be before `AppStore Submission` and `Release` dates?
- The `Code Freeze` date is always before `AppStore Submission`, or those can also be equal?
- The `Release` date is always after `AppStore Submission`, or those can also be equal?
- Should this be validated at all?

-----------------------------

Validating dates is always tricky, I just feel that I should confirm my assumptions before starting anything 😅 

I also feel that I should write a mini-P2 to document those changes, just in case anyone passes through here in the future and needs to do any refactorings. 🗒️

Reply:
Yep writing a P2 seems a good idea.

As for the cases you mention above, those seems to be good ones; I think we should add those, then use `UI.user_error!` in the implementation for when those cases are encountered.

Reply:
> What is the expected behavior of the dates when creating milestones?

Every sprint we:
 - Do the code freeze for version N on Monday, which corresponds to us cutting a `release/*` branch from `trunk`, adding ❄️  to the N milestone, pushing some new adjustments commits in the `release/*` (like freezing the strings to send to translation, bumping the version numbers, …), then create a first beta that we send to beta testers
 - Then we let that period of beta test on the code-frozen beta run for almost a whole sprint (while developers are working on `trunk` on version N+1 in parallel), doing new betas from the `release/*` branch during that period if there are last-minute betafixes to land
 - On the Friday at the end of the Sprint, once the beta-test period is over and we hopefully beta-fixed anything that was urgent to fix during that period, we submit the final build to the App Store / Play Store
 - On the next Monday, providing that Apple/Google have approved the build we submit over the weekend, we start the rollout of that build to end users, and start a new sprint (code freeze of version N+1; wash, rince, repeat)

This process is also described in [this old blog post](https://mobile.blog/2019/05/06/wordpress-mobile-apps-the-heartbeat-release-process/).

Nowadays some of our apps still have 2-weeks sprints, which means when we do a code freeze on Day D+0 (Monday, start of sprint), we then submit for review on Day D+11 (i.e. second Friday), hope it gets approved over the weekend (it usually is), and release on Day D+14 (next Monday), same day we start a new code freeze. Some other apps use 1-week sprints, which means code freeze at D+0, submission at D+4, release at D+7.